test/framework/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregatorTestCase.java

```diff
@@ -133,28 +134,49 @@ public void testUnmappedMissing() throws IOException {
 
     }
 
-    public void testWithSeveralDocs() throws IOException {
+    public void testSingletonDocs() throws IOException {
+        testWithSeveralDocs(() -> true, null);
+    }
+
+    public void testBoundedSingletonDocs() throws IOException {
+        testWithSeveralDocs(() -> true, randomBBox());
+    }
+
+    public void testMultiValuedDocs() throws IOException {
+        testWithSeveralDocs(LuceneTestCase::rarely, null);
+    }
+
+    public void testBoundedMultiValuedDocs() throws IOException {
+        testWithSeveralDocs(LuceneTestCase::rarely, randomBBox());
+    }
+
+    private void testWithSeveralDocs(BooleanSupplier supplier, GeoBoundingBox bbox) throws IOException {
         int precision = randomPrecision();
         int numPoints = randomIntBetween(8, 128);
         Map<String, Integer> expectedCountPerGeoHash = new HashMap<>();
-        testCase(new MatchAllDocsQuery(), FIELD_NAME, precision, null, geoHashGrid -> {
+        testCase(new MatchAllDocsQuery(), FIELD_NAME, precision, bbox, geoHashGrid -> {
             assertEquals(expectedCountPerGeoHash.size(), geoHashGrid.getBuckets().size());
             for (GeoGrid.Bucket bucket : geoHashGrid.getBuckets()) {
                 assertEquals((long) expectedCountPerGeoHash.get(bucket.getKeyAsString()), bucket.getDocCount());
             }
-            assertTrue(AggregationInspectionHelper.hasValue(geoHashGrid));
+            if (bbox == null) {
+                assertTrue(AggregationInspectionHelper.hasValue(geoHashGrid));
+            }
         }, iw -> {
             List<LatLonDocValuesField> points = new ArrayList<>();
             Set<String> distinctHashesPerDoc = new HashSet<>();
             for (int pointId = 0; pointId < numPoints; pointId++) {
                 double[] latLng = randomLatLng();
                 points.add(new LatLonDocValuesField(FIELD_NAME, latLng[0], latLng[1]));
                 String hash = hashAsString(latLng[1], latLng[0], precision);
-                if (distinctHashesPerDoc.contains(hash) == false) {
-                    expectedCountPerGeoHash.put(hash, expectedCountPerGeoHash.getOrDefault(hash, 0) + 1);
+                Rectangle bin = getTile(latLng[1], latLng[0], precision);
+                if (intersectsBounds(bin, bbox) || validPoint(latLng[1], latLng[0], bbox)) {
+                    if (distinctHashesPerDoc.contains(hash) == false) {
+                        expectedCountPerGeoHash.put(hash, expectedCountPerGeoHash.getOrDefault(hash, 0) + 1);
+                    }
+                    distinctHashesPerDoc.add(hash);
                 }
-                distinctHashesPerDoc.add(hash);
-                if (usually()) {
+                if (supplier.getAsBoolean()) {
```

Comment:
we divided the test in two, one that jus return true so it creates singleton documents and one that uses rarely so it will create few multi-value documents.